pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go

```diff
@@ -21,68 +20,74 @@ type AzureTokenProvider interface {
 }
 
 type tokenProviderImpl struct {
-	cfg        *setting.Cfg
-	authParams *plugins.JwtTokenAuth
+	cfg         *setting.Cfg
+	credentials azcredentials.AzureCredentials
+	scopes      []string
 }
 
-func NewAzureAccessTokenProvider(cfg *setting.Cfg, authParams *plugins.JwtTokenAuth) *tokenProviderImpl {
+func NewAzureAccessTokenProvider(cfg *setting.Cfg, credentials azcredentials.AzureCredentials, scopes []string) *tokenProviderImpl {
 	return &tokenProviderImpl{
-		cfg:        cfg,
-		authParams: authParams,
+		cfg:         cfg,
+		credentials: credentials,
+		scopes:      scopes,
 	}
 }
 
 func (provider *tokenProviderImpl) GetAccessToken(ctx context.Context) (string, error) {
 	var credential TokenCredential
 
-	if provider.isManagedIdentityCredential() {
+	// TODO: Move to provider initialization and reuse for each GetAccessToken call
+	switch c := provider.credentials.(type) {
```

Comment:
IMO it's a bit more readable if you use the `AuthType` method for the switch based on a string and do the conversion later on

Reply:
I just used the popular pattern for sum types / discriminated unions:
[here](https://making.pusher.com/alternatives-to-sum-types-in-go/) and [here](https://medium.com/@haya14busa/sum-union-variant-type-in-go-and-static-check-tool-of-switch-case-handling-3bfc61618b1e)

Reply:
in this case `c` is automatically cased to the proper type in each `case` branch